### DISCUSSION TITLE
AJAX error invoking /ajax/dataTypes?... 404

### DISCUSSION THREAD

**Discussion by TaKO8Ki** [phax/phoss-smp#254] [Q&A] [answered]
I deployed SMP and got this error after login. I use a load balancer for this SMP. I'm wondering if it's necessary to configure these properties in application.properties with appropriate values. Is my understanding correct?

```
http.proxyHost = 10.0.0.10
http.proxyPort = 808
https.proxyHost = 10.0.0.10
https.proxyPort = 808
```

<img width="458" alt="Screen Shot 2023-08-31 at 20 54 03" src="https://github.com/phax/phoss-smp/assets/41065217/eeda3fcf-067d-4fad-b43a-b36cfb62d7de">

**Comment by phax** [comment, marked as answer]
If you are using a load balancer, than you should make sure, that all the requests with /ajax/* are always routed to the same instance, because they are based on in-memory data structures of a single instance. You may use the Session Cookie to determine the route I guess....

**Reply by TaKO8Ki** [reply]
Thanks. I will try making the number of instance to 1.

**Reply by TaKO8Ki** [reply]
I have one more question. Does SMP use in-memory data in other implementations other than login management? For instance, when multiple SMP instances are associated with a single load balancer, is it possible to use the SMP REST API successfully? @phax

**Reply by phax** [reply]
If you run multiple instances of the SMP in parallel, you need to use either the SQL or the MongoDB backend. Each of these backends have a user management in the database.

**Reply by TaKO8Ki** [reply]
> If you run multiple instances of the SMP in parallel, you need to use either the SQL or the MongoDB backend. Each of these backends have a user management in the database.

I understand xml backend does not work on multiple instances. I use MySQL backend, so it's ok. In conclusion, if I run multiple instances:

- Login system does not work well on all kinds of backend.
- At least the REST API works correctly if I use SQL or MongoDB backends.

Is my understanding correct?

I will create a pull request for using session cookie to login when I have time.

**Reply by phax** [reply]
* The login system works well on all kind of backends, but the XML backend is not designed for horizontal scaling (multiple instances).
* If you run on the latest version of the SMP, all the credentials are stored in the database. If you run a version prior to v7 I recommend you update to the latest version

**Reply by TaKO8Ki** [reply]
Yes, I'm assuming the use of multiple instances, so I've expressed it as follows:

> if I run multiple instances:
>
> Login system does not work well on all kinds of backend.
> At least the REST API works correctly if I use SQL or MongoDB backends.

My understanding seems right.

> If you run on the latest version of the SMP, all the credentials are stored in the database. If you run a version prior to v7 I recommend you update to the latest version

I already use the latest version, so I think I can use multiple instances at least for REST API, but It's difficult to login with multiple instances. So, I will prepare two load balancers. The first one is for admin users to login and is internal LB associated with one instance. The second one is public LB associated with multiple instances for using REST API.
